pkg/controller/v1alpha1/llmisvc/workload_single_node.go

```diff
@@ -55,3 +215,120 @@ func semanticDeploymentIsEqual(expected *appsv1.Deployment, curr *appsv1.Deploym
+		return err
+	}
+
+	return r.reconcileSingleNodeMainRoleBinding(ctx, llmSvc, serviceAccount, storageConfig)
+}
+
+func (r *LLMISVCReconciler) reconcileSingleNodeMainRole(ctx context.Context, llmSvc *v1alpha1.LLMInferenceService, storageConfig *types.StorageInitializerConfig) error {
+	expectedDeployment, err := r.expectedSingleNodeMainDeployment(ctx, llmSvc, storageConfig)
+	if err != nil {
+		return fmt.Errorf("failed to get expected main deployment: %w", err)
+	}
+
+	role := r.expectedSingleNodeRole(llmSvc)
+	if !hasRoutingSidecar(expectedDeployment.Spec.Template.Spec) {
+		return Delete(ctx, r, llmSvc, role)
+	}
+
+	if err := Reconcile(ctx, r, llmSvc, &rbacv1.Role{}, role, semanticRoleIsEqual); err != nil {
+		return fmt.Errorf("failed to reconcile single node role %s/%s: %w", role.GetNamespace(), role.GetName(), err)
+	}
+
+	return nil
+}
+
+func (r *LLMISVCReconciler) reconcileSingleNodeMainRoleBinding(ctx context.Context, llmSvc *v1alpha1.LLMInferenceService, sa *corev1.ServiceAccount, storageConfig *types.StorageInitializerConfig) error {
+	expectedDeployment, err := r.expectedSingleNodeMainDeployment(ctx, llmSvc, storageConfig)
+	if err != nil {
+		return fmt.Errorf("failed to get expected main deployment: %w", err)
+	}
+
+	roleBinding := r.expectedSingleNodeRoleBinding(llmSvc, sa)
+	if !hasRoutingSidecar(expectedDeployment.Spec.Template.Spec) {
+		return Delete(ctx, r, llmSvc, roleBinding)
+	}
+
+	if err := Reconcile(ctx, r, llmSvc, &rbacv1.RoleBinding{}, roleBinding, semanticRoleBindingIsEqual); err != nil {
+		return fmt.Errorf("failed to reconcile single node rolebinding %s/%s: %w", roleBinding.GetNamespace(), roleBinding.GetName(), err)
+	}
+
+	return nil
+}
+
+func (r *LLMISVCReconciler) expectedSingleNodeMainServiceAccount(llmSvc *v1alpha1.LLMInferenceService) *corev1.ServiceAccount {
+	return &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      kmeta.ChildName(llmSvc.GetName(), "-kserve"),
+			Namespace: llmSvc.GetNamespace(),
+			OwnerReferences: []metav1.OwnerReference{
+				*metav1.NewControllerRef(llmSvc, v1alpha1.LLMInferenceServiceGVK),
+			},
+			Labels: r.singleNodeLabels(llmSvc),
+		},
+	}
+}
+
+func (r *LLMISVCReconciler) expectedSingleNodeRole(llmSvc *v1alpha1.LLMInferenceService) *rbacv1.Role {
+	ro := &rbacv1.Role{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      kmeta.ChildName(llmSvc.GetName(), "-kserve-role"),
+			Namespace: llmSvc.GetNamespace(),
+			OwnerReferences: []metav1.OwnerReference{
+				*metav1.NewControllerRef(llmSvc, v1alpha1.LLMInferenceServiceGVK),
+			},
+			Labels: r.singleNodeLabels(llmSvc),
+		},
+	}
+	ro.Rules = append(ro.Rules, sidecarSSRFProtectionRules...)
+	return ro
+}
+
+func (r *LLMISVCReconciler) expectedSingleNodeRoleBinding(llmSvc *v1alpha1.LLMInferenceService, sa *corev1.ServiceAccount) *rbacv1.RoleBinding {
+	return &rbacv1.RoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      kmeta.ChildName(llmSvc.GetName(), "-kserve-rb"),
```

Comment:
same here, I think we can remove the kserve name to keep the resource name short